Fix delete entry keybindings bug

Braunch opened the pull request:
When using a different localization then English, setting the keybindings does not work correctly (see here #1235 ). I figured out that the problem is that the `KeyEvent` returns localized key texts so you can not use this to set the `KeyBindings`(the JabRef `KeyBindings` use English key texts). At the moment I only see two possibilities that do not end up in comparing tons of Strings to manually get the english localization. 

~~First would be to change the locale to English, so that the `KeyEvent` returns english key texts, but I don't know what I might break with that and if the rest of the localization will still work.~~

 ~~Second solution would be to change the `KeyBindings` so that they use `KeyCodes` instead of Strings. This would IMHO improve the `KeyBindings` system but it would mean a lot of changes in all the `KeyBinding` related code.~~
See the comment below for my approach https://github.com/JabRef/jabref/pull/1815#issuecomment-242346564.

- [x] Manually tested changed features in running JabRef

koppor commented on src/test/resources/testbib/testjabref_292.bib:
```diff
@@ -299,3 +299,4 @@ @siumed.edu
   url = {http://dx.doi.org/10.1016/j.jneumeth.2008.01.009}
 }
 
+@Comment{jabref-meta: databaseType:bibtex;}
```
This is JabRef 3.x - do not change bib files created with old versions of JabRef

Braunch commented:
I now found a way to fix it. I am building a String with the English modifier key text using the triggered KeyEvents  `is-ModifierKey-Down` methods

boceckts commented on CHANGELOG.md:
```diff
@@ -77,6 +77,7 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - Fixed a number of issues related to accessing the GUI from outside the EDT
 - Added a few missing translation strings
 - Fixed NullPointerException when opening Customize entry type dialog without an open database
+- Fixed localization error when changing key bindings while not using the english localization
```
Instead of "Fixed localization error when changing key bindings ..." I would rather write "Fixed modified keybindings not working ..."

boceckts commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,111 @@
+     * @param evt the KeyEvent that was triggered to set the KeyBindings
+     * @return a String containing the modifier keys text
+     */
+    private String getModifierTextAsList(KeyEvent evt) {
```
Rename without "AsList" or "AsListing" as it doesn't return a list.

boceckts commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,111 @@
+                modifiersAsString += " ";
+            }
+            modifiersAsString += s;
+        }
```
This could be replaced by one line by using `modifiersList.stream.collect`

boceckts commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,111 @@
+    private String getModifierTextAsList(KeyEvent evt) {
+        ArrayList<String> modifiersList = new ArrayList<>();
+
+        if (evt.getModifiers() != 0) {
```
Why do you need this? I think `evt.isControlDown()` will return `false` anyway.

boceckts commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,111 @@
+            if (evt.isAltGraphDown()) {
+                modifiersList.add("alt gr");
+            }
+        }
```
I'm missing the meta key, which is especially for mac users an important modifier key. The default shortcuts of JabRef also include this modifier, so please add it.

boceckts commented:
Please resolve conflicts and minor comments, then it's good to me 👍

Braunch commented:
@boceckts can you test it on win and mac for me please? The meta key does not work on my linux system since there are OS functions on it.

boceckts commented:
@Braunch Windows uses the windows key for its own shortcuts, same as with some linux distributions. On mac it now does work to make shortcuts directly with the command(meta) key....but I just realized that if you create a shortcut with the control key, e.g. "ctrl + S" it will automatically be replaced by "command(meta) + S".
So I guess the meta key doesn't have to be listened to..sry for that.

Braunch commented:
Is this intended? Why should the grabbed keys differ from the keys actually used by the user? I changed the behavior so that you can set the 'meta + key' shortcut with the meta. If this is alright I am probably done here.

Braunch commented:
Any opinion on my solution with the meta key from the Developers?

koppor commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,107 @@
 public class KeyBindingsListener extends KeyAdapter {
 
     private final KeyBindingTable table;
+    private boolean isDeleteKey;
```
You don't need to make them class variables, do you?

koppor replied:
Still unanswered. :innocent:

koppor replied:
BUMP

koppor commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,107 @@
+            }
         } else {
-            newKey = modifier.toLowerCase().replace("+", " ") + " " + code;
+            newKey = modifier.toLowerCase(Locale.ENGLISH) + " " + KeyEvent.getKeyText(code);
```
Why is the replacement of "+" to " " no longer needed?

koppor commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,107 @@
+            modifiersList.add("shift");
+        }
+        if (evt.isAltGraphDown()) {
+            modifiersList.add("alt gr");
```
Is that true? altgr without space?

Braunch replied:
On linux 'alt gr' works just fine. On my windows machine at home, 'ctrl alt' is set instead of 'alt gr' but no matter if it's written with or without space in between. I tested it on the master and 'alt gr' is always parsed as 'ctrl alt' so it works as intended I guess.

koppor commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,110 @@
-            boolean isFunctionKey = (kc >= KeyEvent.VK_F1) && (kc <= KeyEvent.VK_F12);
-            boolean isEscapeKey = kc == KeyEvent.VK_ESCAPE;
-            boolean isDeleteKey = kc == KeyEvent.VK_DELETE;
+            isFunctionKey = (kc >= KeyEvent.VK_F1) && (kc <= KeyEvent.VK_F12);
```
why not keeping the delcaration locally? I don't see any usage outside the if statement.

koppor commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,110 @@
+            isEscapeKey = kc == KeyEvent.VK_ESCAPE;
+            isDeleteKey = kc == KeyEvent.VK_DELETE;
             if (!(isFunctionKey || isEscapeKey || isDeleteKey)) {
                 return; // need a modifier except for function, escape and delete keys
```
Move comment form after the return to the line above

koppor replied:
BUMP

koppor commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,110 @@
+        if (code == KeyEvent.VK_ALT ||
+                code == KeyEvent.VK_TAB ||
+                code == KeyEvent.VK_BACK_SPACE ||
+                code == KeyEvent.VK_ENTER ||
```
Could you comment why `ENTER` is treated as modifier? Also applies to "TAB" and "BACK_SPACE". Maybe the reason is that these keys should not be used at shortcuts at all? Maybe a new `if` statement should be introduced with an appropriate comment.

koppor replied:
BUMP

koppor commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBindingsListener.java:
```diff
@@ -2,65 +2,110 @@
     }
+
+    /**
+     * Checks if the event has pressed modifiers and adds their English key text to an ArrayList
```
The internal implementation should not exposed. Formulate it in an abstract way:

> Collects the English translations of all modifiers and returns a space separated list

koppor replied:
BUMP

koppor commented:
Please merge `upstream/mater` and fix minor comments. Then, it should be good to go.

Braunch commented:
I addressed the small comments and resolved the conflicts.

koppor commented:
You commited with both your st university address and your normal development address. Please double check your git config.

Please merge upstream/master again. Can you make one commit as described at https://github.com/JabRef/jabref/wiki/Tools#rebase-everything-as-one-commit-on-master? Then, I can resolve the conflicts by myself.

Braunch commented:
I resolved the conflicts and squashed my commits. Should be good to go.

koppor commented:
Please do not squash, this makes reviewing very hard as we cannot see, what has been adressed. I added a `BUMP` on all my unaswered comments. Please fix/answer

Braunch commented:
I addressed your comments and merged the upstream/master. I squashed following your comment https://github.com/JabRef/jabref/pull/1815#issuecomment-246499706 Did I get that wrong?

Braunch commented on CHANGELOG.md:
```diff
@@ -45,12 +45,13 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - Entries in the SearchResultDialog are now converted to Unicode
 - Suggestions in the autocomplete (search) are now in Unicode
-- Fixed NullPointerException when opening search result window for an untitled database 
+- Fixed NullPointerException when opening search result window for an untitled database
```
I encountered an unnecessary space so i deleted it
